crates/bevy_ecs/src/reflect.rs

```diff
@@ -110,14 +117,17 @@ impl ReflectComponent {
     }
 }
 
-impl<C: Component + Reflect + FromWorld> FromType<C> for ReflectComponent {
+impl<C: Component + Reflect + FromReflect> FromType<C> for ReflectComponent {
     fn from_type() -> Self {
         ReflectComponent {
             insert: |world, entity, reflected_component| {
-                let mut component = C::from_world(world);
-                component.apply(reflected_component);
+                let component = C::from_reflect(reflected_component).unwrap();
```

Comment:
Can we make this and all other `unwrap` calls (where necessary) `expect` calls instead? There are times when this might fail, and a good error message would be nice 🙂 

Examples of when this could go wrong:
* A user manually implemented `FromReflect` themselves and did so in a way that could return `None`
* The wrong `reflected_component` was passed (happened to me before and is a nightmare to debug). This can happen when an invalid `DynamicStruct` is passed.

Reply:
Just swapped a bunch with `expect`, how are the error messages? I'd make them more descriptive but I'm unaware of any `.expect_with()` method to let me do that without spending time constructing a formatted string on every operation.

Reply:
Hm, I might be more specific and say something like `"failed to call FromReflect::from_reflect on dynamic component"` or something like that. Just to make it clear that it's `FromReflect` where the issue is. Thoughts on that?

Reply:
SGTM!